### PR TITLE
feat: push lam image to the embedded cluster registry

### PR DESCRIPTION
**Pull request by ricardomaraschini**
#### What this PR does / why we need it:

after pushing application image we also need to push all embedded cluster images. at this stage only the lam image is shipped.

validation after install:

```
$ docker pull ec:5000/embedded-cluster/embedded-cluster-local-artifact-mirror:dev-54f8048
dev-54f8048: Pulling from embedded-cluster/embedded-cluster-local-artifact-mirror
aae52d44a2a5: Already exists
32085ade9ebb: Pull complete
Digest: sha256:a2745ce1e8605b926de73187c6112f52acf31167fc0c0b0747fc9a3b2904f4ab
Status: Downloaded newer image for ec:5000/embedded-cluster/embedded-cluster-local-artifact-mirror:dev-54f8048
ec:5000/embedded-cluster/embedded-cluster-local-artifact-mirror:dev-54f8048
$
```

#### Which issue(s) this PR fixes:
Relates #105498

#### Does this PR introduce a user-facing change?
```release-note
Pushes embedded cluster images to the registry during install time
```

#### Does this PR require documentation?
NONE

### CODE REVIEW

File: pkg/image/airgap.go [hunk 1]
```diff
@@ -187,8 +190,23 @@ func TagAndPushImagesFromBundle(airgapBundle string, options imagetypes.PushImag
...
+	return PushImagesUsingTempRegistry(imagesDir, artifacts.Registry.SavedImages, options, statusMessage)
+}
+
+func PushAppImagesFromTempRegistry(airgapRootDir string, airgap *kotsv1beta1.Airgap, options imagetypes.PushImagesOptions) error {
 	imagesDir := filepath.Join(airgapRootDir, "images")
+	statusMessage := "Pushing application images (%d/%d)\n"
+	return PushImagesUsingTempRegistry(imagesDir, airgap.Spec.SavedImages, options, statusMessage)
+}
+
+func PushImagesUsingTempRegistry(imagesDir string, imageList []string, options imagetypes.PushImagesOptions, statusMessage string) error {
```
Comment (sgalsaleh) [inline pkg/image/airgap.go:209]: nit pick: 
```suggestion
func PushImagesFromTempRegistry(imagesDir string, imageList []string, options imagetypes.PushImagesOptions, statusMessage string) error {
```

File: pkg/image/airgap.go [hunk 2]
```diff
@@ -187,6 +191,104 @@ func TagAndPushImagesFromBundle(airgapBundle string, options imagetypes.PushImag
...
+		imageParts := strings.Split(srcImage, "/")
+		lastPart := imageParts[len(imageParts)-1]
+		destStr := fmt.Sprintf("docker://%s/embedded-cluster/%s", options.Registry.Endpoint, lastPart)
+		destRef, err := alltransports.ParseImageName(destStr)
+		if err != nil {
+			return errors.Wrapf(err, "failed to parse dest image %s", destStr)
+		}
+
+		// copy all architecures available in the bundle.
+		// this also handles kotsadm airgap bundles that have multi-arch images but are referenced by tag.
```
Comment (sgalsaleh) [inline pkg/image/airgap.go:258]: Let's remove this line about kotsadm as it doesn't apply here.

File: pkg/image/airgap.go [hunk 3]
```diff
@@ -187,6 +191,105 @@ func TagAndPushImagesFromBundle(airgapBundle string, options imagetypes.PushImag
...
 
+func PathToRegistryECImage(srcImage, registryNamespace string) string {
+	imageParts := strings.Split(srcImage, "/")
+	imageName := imageParts[len(imageParts)-1]
+	return filepath.Join(registryNamespace, "embedded-cluster", imageName)
+}
+
+func PushECImagesFromTempRegistry(airgapRootDir string, airgap *kotsv1beta1.Airgap, options imagetypes.PushImagesOptions) error {
+	artifacts := airgap.Spec.EmbeddedClusterArtifacts
+	if artifacts.Registry.Dir == "" || len(artifacts.Registry.SavedImages) == 0 {
```
Comment (sgalsaleh) [inline pkg/image/airgap.go:202]: Let's also check if "artifacts" is nil?

File: pkg/image/airgap.go [hunk 4]
```diff
@@ -187,6 +191,105 @@ func TagAndPushImagesFromBundle(airgapBundle string, options imagetypes.PushImag
 	return nil
 }
 
+func PathToRegistryECImage(srcImage, registryNamespace string) string {
```
Comment (sgalsaleh) [inline pkg/image/airgap.go:194]: Can you please add a unit test for this?

File: pkg/image/airgap.go [hunk 5]
```diff
@@ -187,6 +191,105 @@ func TagAndPushImagesFromBundle(airgapBundle string, options imagetypes.PushImag
...
+	imageParts := strings.Split(srcImage, "/")
+	imageName := imageParts[len(imageParts)-1]
+	return filepath.Join(registryNamespace, "embedded-cluster", imageName)
+}
+
+func PushECImagesFromTempRegistry(airgapRootDir string, airgap *kotsv1beta1.Airgap, options imagetypes.PushImagesOptions) error {
```
Comment (sgalsaleh) [inline pkg/image/airgap.go:200]: How much effort would jt be to add a unit test for this function similar to the test we have for pushing embedded cluster artifacts?

Comment (ricardomaraschini) [reply pkg/image/airgap.go:200]: That would consume me half a day, at the very least. The test is much more closer to PushImagesFromTempRegistry than to the artifacts function. Let me know if I should start it.

**Review by sgalsaleh — APPROVED** [review]

File: pkg/image/airgap.go (hunk 5 above)
Comment (sgalsaleh) [reply pkg/image/airgap.go:203]: No need, we can cover it in the e2e test. Thank you.

**Pull request merged**